beater/pub.go

```diff
@@ -52,44 +61,47 @@ func newPublisher(pipeline beat.Pipeline, N int, shutdownTimeout time.Duration)
 
 		// Set channel size to N - 1. One request will be actively processed by the
 		// worker, while the other concurrent requests will be buffered in the queue.
-		events: make(chan []beat.Event, N-1),
+		pendingRequests: make(chan pendingReq, N-1),
+	}
+
+	for i := 0; i < runtime.GOMAXPROCS(0); i++ {
+		go p.run()
 	}
 
-	go p.run()
 	return p, nil
 }
 
 // Stop closes all channels and waits for the the worker to stop.
-// The worker will drain the queue on shutdown, but no more events
+// The worker will drain the queue on shutdown, but no more pending requests
 // will be published.
 func (p *publisher) Stop() {
 	p.m.Lock()
 	p.stopped = true
 	p.m.Unlock()
-	close(p.events)
+	close(p.pendingRequests)
 	p.client.Close()
 }
 
-// Send tries to forward events to the publishers worker. If the queue is full,
+// Send tries to forward pendingReq to the publishers worker. If the queue is full,
 // an error is returned.
 // Calling send after Stop will return an error.
-func (p *publisher) Send(batch []beat.Event) error {
+func (p *publisher) Send(req pendingReq) error {
 	p.m.RLock()
 	defer p.m.RUnlock()
 	if p.stopped {
-		return errChanneClosed
+		return errChannelClosed
 	}
 
 	select {
-	case p.events <- batch:
+	case p.pendingRequests <- req:
 		return nil
 	case <-time.After(time.Second * 1): // this forces the go scheduler to try something else for a while
 		return errFull
 	}
 }
 
 func (p *publisher) run() {
-	for batch := range p.events {
-		p.client.PublishAll(batch)
+	for req := range p.pendingRequests {
+		p.client.PublishAll(req.payload.Transform(req.config))
```

Comment:
could we put the config as property on the publisher so we didn't have to pass it in every `pendingReq` ? It doesn't change know

Reply:
https://github.com/elastic/apm-server/pull/746#discussion_r179128363

Actually I also think it is cleaner to separate the config from the payload data.

Reply:
Oh sorry, you said on the publisher, I misread. We cannot put it on the publisher, as it is different per endpoint.